lib/Monad_WP/OptionMonad.thy

```diff
@@ -62,6 +62,30 @@ lemma opt_map_upd_Some:
 
 lemmas opt_map_upd[simp] = opt_map_upd_None opt_map_upd_Some
 
+lemma opt_map_Some_comp[simp]:
+  "f ||> h o g = f ||> g ||> h"
+  by (fastforce simp: opt_map_def split: option.split)
+
+lemma opt_map_fold_l:
+  "(id |> g) (f x) = (f |> g) x"
+  by (clarsimp simp: opt_map_def)
+
+(* LHS is the same as (f ||> id) *)
+lemma opt_map_unit_r[simp]:
+  "(f |> Some) x = f x"
```

Comment:
Sorry for spotting this only now, but it would make sense to eta-contract this one to `f |> Some = f`. This would mean it can apply in the middle of a chain as well.

Reply:
Ah, yes, that makes sense. I was hoping these rules to be something of algebra-like, too.